documentation/Architecture/Overview.md

```diff
@@ -2,80 +2,105 @@
+4) Processing (also known as cooking) data within a plugin through a [processing pipeline](./The-Data-Processing-Pipeline)
+5) Querying and piping data between either two concurrently loaded plugins or a loaded plugin and an assembly with programmatic 
+   access to the running instance of the SDK (even if the data originates from a plugin which the destination does 
+   not have the source code for)
 
-In the above example, there must be an ETL plugin to read an .etl trace file (data source).
+## Plugins
 
+From an interface perspective, a plugin is a collection of
+* One or more `CustomDataSource`s
+* One `CustomDataProcessor` for each `CustomDataSource`
+* Zero or more `Table`s
+* Zero or more `DataCooker`s
 
-## Custom Data Source Model
-<details>
+that work together to
+1) Consume specific data sources that the SDK provides
+2) Create populated tables that get sent to the SDK
+3) Expose data that can be programmatically accessed by outside sources, such as other plugins
 
-<summary>Click to Expand</summary>
- 
+A plugin is an **abstract collection of these objects** which can be bundled together for distribution and loaded as a single 
+unit by the SDK driver. *A plugin can be, and often is, made up of multiple assemblies*.
 
-A Custom Data Source (CDS) is a containerized unit which has instructions to parse the Data Source. 
-Every Custom Data Source has to advertise the supported file types (Data Source). 
-If the relevant binaries exist, the CDS has the logic for creating tables from the Data Source.
+:warning: Note that while a single assembly *can* define more than one `CustomDataSource`, __it is highly recommended that an assembly only contains 
+a single `CustomDataSource`.__ Tables, data cookers, and custom data processors are almost always associated with a single `CustomDataSource`. 
+It is best therefore to package __only one__ `CustomDataSource` and all of its associated classes in a single binary. 
 
-The tables are returned as binary instructions from the Data Processor to the SDK to create 0 or more tables.
-The Driver (WPA is recommended) dynamically loads each Custom Data Source at runtime through the SDK.
+The diagram below demonstrates how these objects work together to acheive this high-level interface:
 
-The Custom Data Source Model allows developers to use any arbitrary Data Source with the SDK to build desired tables using plugins. 
+<img src=".attachments/plugin.svg" width="800">
 
-</details>
+1) The `CustomDataSource` advertises to the SDK which data sources it supports and all the tables it can output. The `CustomDataSource` is 
+   primarily an interface definition, but it also implements creating a `CustomDataProcessor`
+2) The `CustomDataProcessor` contains implementation logic for processing the data sources provided by the SDK. The data sources 
+   are given to the `CustomDataSource`, which typically passes them to the `CustomDataProcessor` it creates
```

Comment:
This sentence is a bit hard for me. The last portion "it create" seems unnecessary or make the sentence active?

Reply:
I agree. This sentence is kind of awkward because the subject switches from the CDP to the data sources between the first and second sentence. How would this sound:

2) The `CustomDataProcessor` implements logic for processing the data sources provided to the `CustomDataSource`. It learns about the data sources it must process from the `CustomDataSource` that creates it.

Or do you still think that last sentence is unncessary?